packages/shopping-cart/src/shopping-cart-actions.ts

```diff
@@ -0,0 +1,56 @@
+import { createRequestCartProducts } from './create-request-cart-product';
+import type {
+	AddProductsToCart,
+	ReplaceProductsInCart,
+	RemoveProductFromCart,
+	ReplaceProductInCart,
+	UpdateTaxLocationInCart,
+	ApplyCouponToCart,
+	RemoveCouponFromCart,
+	ReloadCartFromServer,
+	RequestCartProduct,
+	DispatchAndWaitForValid,
+	ShoppingCartActionCreators,
+} from './types';
+
+export function createActionCreators(
+	dispatch: DispatchAndWaitForValid
+): ShoppingCartActionCreators {
+	const removeCoupon: RemoveCouponFromCart = () => dispatch( { type: 'REMOVE_COUPON' } );
+	const addProductsToCart: AddProductsToCart = ( products ) =>
+		dispatch( {
+			type: 'CART_PRODUCTS_ADD',
+			products: createRequestCartProducts( products ),
+		} );
+	const removeProductFromCart: RemoveProductFromCart = ( uuidToRemove ) =>
+		dispatch( { type: 'REMOVE_CART_ITEM', uuidToRemove } );
+	const replaceProductsInCart: ReplaceProductsInCart = ( products ) =>
+		dispatch( {
+			type: 'CART_PRODUCTS_REPLACE_ALL',
+			products: createRequestCartProducts( products ),
+		} );
+	const replaceProductInCart: ReplaceProductInCart = (
+		uuidToReplace: string,
+		productPropertiesToChange: Partial< RequestCartProduct >
+	) =>
```

Comment:
When written this way, you need to duplicate the method type definitions that are already defined on `ShoppingCartActionCreators` and could be inferred from there. A more concise way to write this would be:
```js
return {
  replaceProductInCart: ( uuidToReplace, productPropertiesToChange ) => dispatch( ... ),
  ...
}
```
Here TypeScript knows that it's creating an instance of `ShoppingCartActionCreators` and it can infer all the param and return types of `replaceProductInCart` -- we don't need to specify anything at all!

TypeScript is capable of inferring many types, all it needs is just a little help from its programmers.